include/mbedtls/x509_crt.h

```diff
@@ -675,6 +746,232 @@ void mbedtls_x509_crt_restart_init( mbedtls_x509_crt_restart_ctx *ctx );
+                                        mbedtls_x509_sequence **ext_key_usage );
+
+/**
+ * \brief           Flush internal X.509 CRT parsing cache, if present.
+ *
+ * \param crt       The CRT structure whose cache to flush.
+ *
+ * \note            Calling this function frequently reduces RAM usage
+ *                  at the cost of performance.
+ *
+ * \return          \c 0 on success.
+ * \return          A negative error code on failure.
+ */
+int mbedtls_x509_crt_flush_cache( mbedtls_x509_crt const *crt );
+
+/* Internal X.509 CRT cache handling functions.
+ * They are not part of the public API and may change
+ * at any time. */
+
+int mbedtls_x509_crt_flush_cache_frame( mbedtls_x509_crt const *crt );
+int mbedtls_x509_crt_flush_cache_pk( mbedtls_x509_crt const *crt );
+
+int mbedtls_x509_crt_cache_provide_frame( mbedtls_x509_crt const *crt );
+int mbedtls_x509_crt_cache_provide_pk( mbedtls_x509_crt const *crt );
+
+static inline int mbedtls_x509_crt_frame_acquire( mbedtls_x509_crt const *crt,
+                                                  mbedtls_x509_crt_frame **frame_ptr )
+{
+    int ret;
+#if defined(MBEDTLS_THREADING_C)
+    if( mbedtls_mutex_lock( &crt->cache->frame_mutex ) != 0 )
+        return( MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+#endif
+
+    ret = mbedtls_x509_crt_cache_provide_frame( crt );
+    if( ret != 0 )
+    {
+#if defined(MBEDTLS_THREADING_C)
+        if( mbedtls_mutex_unlock( &crt->cache->frame_mutex ) != 0 )
+            return( MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+#endif
+        return( ret );
+    }
+
+    *frame_ptr = crt->cache->frame;
+    return( 0 );
+}
+
+static inline void mbedtls_x509_crt_frame_release( mbedtls_x509_crt const *crt )
+{
+    ((void) crt);
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_unlock( &crt->cache->frame_mutex );
+#endif
+
+#if defined(MBEDTLS_X509_ALWAYS_FLUSH)
+    (void) mbedtls_x509_crt_flush_cache_frame( crt );
+#endif /* MBEDTLS_X509_ALWAYS_FLUSH */
+}
+
+static inline int mbedtls_x509_crt_pk_acquire( mbedtls_x509_crt const *crt,
+                                               mbedtls_pk_context **pk_ptr )
+{
+    int ret;
+#if defined(MBEDTLS_THREADING_C)
+    if( mbedtls_mutex_lock( &crt->cache->pk_mutex ) != 0 )
+        return( MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+#endif
+
+    ret = mbedtls_x509_crt_cache_provide_pk( crt );
+    if( ret != 0 )
+    {
+#if defined(MBEDTLS_THREADING_C)
+        if( mbedtls_mutex_unlock( &crt->cache->pk_mutex ) != 0 )
+            return( MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+#endif
+        return( ret );
+    }
+
+    *pk_ptr = crt->cache->pk;
+    return( 0 );
+}
+
+static inline void mbedtls_x509_crt_pk_release( mbedtls_x509_crt const *crt )
+{
+    ((void) crt);
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_unlock( &crt->cache->pk_mutex );
+#endif
+
+#if defined(MBEDTLS_X509_ALWAYS_FLUSH)
+    (void) mbedtls_x509_crt_flush_cache_pk( crt );
+#endif /* MBEDTLS_X509_ALWAYS_FLUSH */
+}
```

Comment:
In large systems with many concurrent connections accessing the same CRT, you would probably not flush caches, so no re-parsing will be done. However, the current implementation would still use locks while a CRT frame or PK context is in use, thereby degrading performance - I agree.

Is this avoidable? Logically, it seems so, because each thread should use CRT frame and PK context in a read-only manner, and therefore one should get away with some kind of reader/writer lock.

For CRT frames, I indeed don't see a problem with changing the signature of `mbedtls_x509_crt_frame_acquire()` to
```c
static inline int mbedtls_x509_crt_frame_acquire( mbedtls_x509_crt const *crt,
                                                  mbedtls_x509_crt_frame const **frame_ptr )
```

For the PK context, however, I don't think we can do that at the moment, because, unfortunately, signing and verification operations aren't `const` on PK contexts.

We could, however, keep the signature of `mbedtls_x509_crt_pk_acquire()` and explicitly document that the resulting PK context must only be used for operations which are known to be thread-safe, which includes all implementations of sign/verify operations that we currently provide. (Note that this limitation is already implicit in `mbedtls_x509_crt::pk` as we have it so far, as the trusted CAs are part of `mbedtls_ssl_config` and hence shared and concurrently used by multiple threads using the same config). This would allow to use a reader/writer lock in the implementation to lock concurrent attempts to fill/free a cache, but allow concurrent uses of a filled cache. 

What do you think?

Reply:
Any comments wrt this?

Reply:
> In other words, if I want to open 10 TLS connections then I potentially need to parse some components of my CA chain 10 times. To mitigate this, there is another struct that allows caching some of this information, such as the PK.

If you use a cache, you need to reparse only those parts of the CRT that have previously been stored in dynamically allocated linked lists: subject, issuer, ExtKeyUsage, SAN. Everything else is kept and need not be reparsed. It's not clear to me that memory-local re-parsing these components is necessarily slower than than traversing a linked list that's scattered in memory.

If you don't use a cache, then of course it's slower, but you wouldn't do that in a server context.

> Unfortunately, computing the data in the cache isnt reentrant, so we also need these locks to access the cached data.

"Yes" for the first part, "No" for the second: Creating and flushing the cache must be locked, but reading it need not be. That's what I meant earlier: I agree that the earlier version, which locked the reading, was inefficient, but the new version, which only locks creating/flushing the cache, isn't.

I have added a test which exercises concurrent X.509 CRT verification using pthread, and on my machine it shows no significant difference in speed, as expected. The test runs 100 threads in parallel each verifying the same pair of CRT/CA 10 times.

```
!MBEDTLS_X509_ON_DEMAND_PARSING + MBEDTLS_X509_ALWAYS_FLUSH:

% for i in `seq 1 20`; do time ( ./test_suite_x509parse_pthread > /dev/null ); done
( ./test_suite_x509parse_pthread > /dev/null; )  14.69s user 0.07s system 692% cpu 2.132 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.41s user 0.06s system 730% cpu 1.982 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.40s user 0.04s system 714% cpu 2.020 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.55s user 0.07s system 715% cpu 2.042 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.45s user 0.04s system 730% cpu 1.982 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.53s user 0.04s system 716% cpu 2.035 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.56s user 0.02s system 718% cpu 2.029 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.43s user 0.03s system 733% cpu 1.973 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.32s user 0.05s system 730% cpu 1.968 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.54s user 0.04s system 718% cpu 2.028 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.30s user 0.05s system 732% cpu 1.960 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.31s user 0.08s system 729% cpu 1.973 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.34s user 0.05s system 732% cpu 1.965 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.56s user 0.04s system 725% cpu 2.015 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.56s user 0.04s system 717% cpu 2.036 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.54s user 0.06s system 705% cpu 2.070 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.37s user 0.04s system 733% cpu 1.964 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.54s user 0.04s system 714% cpu 2.041 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.64s user 0.04s system 706% cpu 2.078 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.47s user 0.06s system 733% cpu 1.981 total

!MBEDTLS_X509_ON_DEMAND_PARSING + !MBEDTLS_X509_ALWAYS_FLUSH:
% for i in `seq 1 20`; do time ( ./test_suite_x509parse_pthread > /dev/null ); done
( ./test_suite_x509parse_pthread > /dev/null; )  14.75s user 0.07s system 692% cpu 2.138 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.59s user 0.02s system 735% cpu 1.987 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.54s user 0.03s system 733% cpu 1.986 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.59s user 0.06s system 733% cpu 1.997 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.53s user 0.03s system 734% cpu 1.981 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.48s user 0.03s system 732% cpu 1.980 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.53s user 0.05s system 733% cpu 1.989 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.51s user 0.04s system 732% cpu 1.986 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.59s user 0.06s system 732% cpu 2.001 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.67s user 0.04s system 733% cpu 2.005 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.74s user 0.04s system 734% cpu 2.014 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.46s user 0.03s system 734% cpu 1.973 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.62s user 0.08s system 733% cpu 2.004 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.93s user 0.02s system 726% cpu 2.059 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.66s user 0.05s system 706% cpu 2.082 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.67s user 0.03s system 715% cpu 2.055 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.52s user 0.02s system 734% cpu 1.979 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.50s user 0.02s system 734% cpu 1.978 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.52s user 0.04s system 733% cpu 1.986 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.78s user 0.04s system 735% cpu 2.014 total

MBEDTLS_X509_ON_DEMAND_PARSING + MBEDTLS_X509_ALWAYS_FLUSH:

for i in `seq 1 20`; do time ( ./test_suite_x509parse_pthread > /dev/null ); done
( ./test_suite_x509parse_pthread > /dev/null; )  14.84s user 0.05s system 689% cpu 2.161 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.59s user 0.02s system 730% cpu 1.999 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.77s user 0.05s system 699% cpu 2.120 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.64s user 0.05s system 701% cpu 2.095 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.60s user 0.06s system 731% cpu 2.004 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.72s user 0.05s system 731% cpu 2.018 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.64s user 0.05s system 731% cpu 2.010 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.68s user 0.06s system 710% cpu 2.076 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.63s user 0.04s system 732% cpu 2.003 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.76s user 0.05s system 704% cpu 2.102 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.87s user 0.04s system 730% cpu 2.040 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.67s user 0.04s system 697% cpu 2.110 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.79s user 0.04s system 707% cpu 2.096 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.77s user 0.06s system 704% cpu 2.106 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.78s user 0.05s system 710% cpu 2.086 total
( ./test_suite_x509parse_pthread > /dev/null; )  15.01s user 0.05s system 727% cpu 2.068 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.77s user 0.06s system 703% cpu 2.108 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.53s user 0.07s system 730% cpu 1.998 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.82s user 0.04s system 731% cpu 2.030 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.71s user 0.07s system 711% cpu 2.077 total

MBEDTLS_X509_ON_DEMAND_PARSING + !MBEDTLS_X509_ALWAYS_FLUSH:
for i in `seq 1 20`; do time ( ./test_suite_x509parse_pthread > /dev/null ); done
( ./test_suite_x509parse_pthread > /dev/null; )  14.79s user 0.06s system 694% cpu 2.136 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.55s user 0.03s system 735% cpu 1.982 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.55s user 0.04s system 732% cpu 1.991 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.56s user 0.03s system 733% cpu 1.990 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.68s user 0.02s system 735% cpu 1.999 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.69s user 0.03s system 734% cpu 2.004 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.54s user 0.04s system 735% cpu 1.984 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.73s user 0.05s system 720% cpu 2.052 total
( ./test_suite_x509parse_pthread > /dev/null; )  15.00s user 0.06s system 708% cpu 2.126 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.55s user 0.03s system 734% cpu 1.984 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.60s user 0.04s system 732% cpu 1.997 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.73s user 0.04s system 725% cpu 2.037 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.42s user 0.07s system 733% cpu 1.975 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.64s user 0.06s system 707% cpu 2.077 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.56s user 0.03s system 732% cpu 1.993 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.57s user 0.02s system 735% cpu 1.985 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.55s user 0.02s system 734% cpu 1.984 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.58s user 0.02s system 734% cpu 1.989 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.52s user 0.03s system 733% cpu 1.983 total
( ./test_suite_x509parse_pthread > /dev/null; )  14.74s user 0.04s system 733% cpu 2.016 total
```

(However, with the old version of cache locking, which didn't allow concurrent reading, one gets:
```
for i in `seq 1 20`; do time ( ./test_suite_x509parse_pthread > /dev/null ); done
( ./test_suite_x509parse_pthread > /dev/null; )  7.68s user 0.06s system 101% cpu 7.633 total
( ./test_suite_x509parse_pthread > /dev/null; )  7.55s user 0.02s system 101% cpu 7.481 total
( ./test_suite_x509parse_pthread > /dev/null; )  7.56s user 0.03s system 101% cpu 7.494 total
( ./test_suite_x509parse_pthread > /dev/null; )  7.56s user 0.04s system 101% cpu 7.503 total
( ./test_suite_x509parse_pthread > /dev/null; )  7.56s user 0.04s system 101% cpu 7.511 total
```
hence a significant performance penalty).

Final note: It is expected that in a highly parallelized setup, `MBEDTLS_X509_ALWAYS_FLUSH` doesn't make a big difference, because it can only flush the cache if it's unused, which is rare if 100 threads use it concurrently. However, even if one just has a single thread verifying a CRT over and over again, there's doesn't seem to be a difference, which I find a bit surprising, even though it's clear that the parsing is nowhere as expensive as the asymmetric crypto.